torch/csrc/jit/codegen/cuda/executor.cpp

```diff
@@ -283,42 +285,111 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
+    launch_params = executor_entry->launch_params;
+    for (size_t i = 0; i < executor_entry->output_sizes.size(); i++) {
+      auto tensor_options = at::TensorOptions()
+                                .dtype(executor_entry->output_types[i])
+                                .device(options_.device);
+      alloced_outputs.push_back(
+          at::empty(executor_entry->output_sizes[i], tensor_options));
+    }
+    for (size_t i = 0; i < executor_entry->empty_buffer_sizes.size(); i++) {
+      auto tensor_options = at::TensorOptions()
+                                .dtype(executor_entry->empty_buffer_types[i])
+                                .device(options_.device);
+      global_buffers.empty_buffers.push_back(
+          at::empty(executor_entry->empty_buffer_sizes[i], tensor_options));
+    }
+    for (size_t i = 0; i < executor_entry->zero_buffer_sizes.size(); i++) {
+      auto tensor_options = at::TensorOptions()
+                                .dtype(executor_entry->zero_buffer_types[i])
+                                .device(options_.device);
+      global_buffers.zero_buffers.push_back(
+          at::zeros(executor_entry->zero_buffer_sizes[i], tensor_options));
+    }
+    rand_offset = executor_entry->rand_offset;
+  } else {
+    executor_utils::validateKernelInputs(&fusion_, inputs, options_.device);
+
+    EvaluationContext evaluation_context =
+        executor_utils::bindInputs(inputs, &fusion_, &lowered_);
 
-  LaunchParams launch_params =
-      computeLaunchParams(inputs, launch_constraints, evaluation_context);
+    launch_params =
+        computeLaunchParams(inputs, launch_constraints, evaluation_context);
 
-  std::vector<at::Tensor> alloced_outputs = outputs;
-  if (outputs.empty() || outputs.size() != fusion_.outputs().size()) {
-    alloced_outputs = allocOutputs(evaluation_context);
-  }
+    if (outputs.empty() || outputs.size() != fusion_.outputs().size()) {
+      alloced_outputs = allocOutputs(evaluation_context);
+    }
 
-  executor_utils::validateKernelOutputs(
-      &fusion_, alloced_outputs, options_.device);
+    executor_utils::validateKernelOutputs(
+        &fusion_, alloced_outputs, options_.device);
+
+    global_buffers = allocGlobalVals(evaluation_context);
+
+    if (has_random_) {
+      // NOTE: this is how we map offset to PW kernels in order to have
+      // identical random number generator to match native PyTorch results.
+      // But it doesn't really work as it takes assumption how threads are
+      // binded but is not generally how we handle that in scheduler.
+      // Refer to `Philox` in generated kernel to understand how the mapping
+      // works.
+      rand_offset = 4 *
+          (std::ceil(
+               alloced_outputs[0].numel() /
+               (4.0 * 128 * launch_params.gdimx())) + // NOLINT
+           1);
+    }
+
+    // record the the short-cut executor entry for the given input set;
+    if (executor_entry) {
```

Comment:
Isn't executor_entry nullptr if we didn't have an entry? Shouldn't we be creating a new entry here if we didn't have one? Maybe I wasn't reading carefully enough.

Reply:
It looks like maybe you're always creating an entry at a higher level in the stack, assuming execution goes through JIT.

Reply:
Are you resetting these values below even if we already have filled out values?

Reply:
This one is quite confusing. It is to address our testing API where runFusion is used directly without an encoding for short-cut. At the beginning we have:
```
  ExecutorEntry* executor_entry = nullptr;
  if (opt_code.has_value()) {
    executor_entry = &executor_entry_lookup_[*opt_code];
  }
```

Reply:
Basically this skips the short-cut recording when we do not provide a key to the short-cut map, in which case there's no entry provided.

Reply:
Note to myself that it would be beneficial to have code path & stages in comments for better readability.